# HW 8-1 
這次還是用回上次的資料庫來訓練

In [83]:
%env KERAS_BACKEND=tensorflow

%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt

from keras.datasets import mnist
from keras.utils import np_utils

from keras.models import Sequential
from keras.layers import Dense, Activation
from keras.optimizers import SGD

from keras.models import Model
from keras.layers import Input

from keras.layers import concatenate, add

env: KERAS_BACKEND=tensorflow


In [0]:
(x_train_1, y_train_1), (x_test_1, y_test_1) = mnist.load_data()

In [0]:
x_train_1 = x_train_1.reshape(60000, 784)
x_test_1 = x_test_1.reshape(10000, 784)

In [0]:
y_train_1 = np_utils.to_categorical(y_train_1, 10)
y_test_1 = np_utils.to_categorical(y_test_1, 10)

In [0]:
f_1 = Dense(550, activation='sigmoid')
f_2 = Dense(150, activation='relu')
f_3 = Dense(50, activation='sigmoid')
f_4 = Dense(50, activation='relu')
f_5 = Dense(10, activation='softmax')

In [0]:
x = Input(shape=(784,))

In [0]:
h_1 = f_1(x)
h_2 = f_2(h_1)
h_3 = f_3(h_2)
h_4 = f_4(h_2)

In [0]:
u = concatenate([h_3, h_4])
y = f_5(u)

In [91]:
model = Model(x, y)
model.summary()

__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_9 (InputLayer)            (None, 784)          0                                            
__________________________________________________________________________________________________
dense_41 (Dense)                (None, 550)          431750      input_9[0][0]                    
__________________________________________________________________________________________________
dense_42 (Dense)                (None, 150)          82650       dense_41[0][0]                   
__________________________________________________________________________________________________
dense_43 (Dense)                (None, 50)           7550        dense_42[0][0]                   
__________________________________________________________________________________________________
dense_44 (

In [0]:
model.compile(loss='mse', optimizer=SGD(lr=0.08), metrics=['accuracy'])

In [93]:
model.fit(x_train_1, y_train_1, 
          batch_size=60, 
          epochs=25, 
          verbose=1, 
          validation_data=(x_test_1, y_test_1))

Train on 60000 samples, validate on 10000 samples
Epoch 1/25
60000/60000 [==============================] - 9s 155us/step - loss: 0.0500 - acc: 0.6467 - val_loss: 0.0219 - val_acc: 0.8733
Epoch 2/25
60000/60000 [==============================] - 9s 143us/step - loss: 0.0175 - acc: 0.8959 - val_loss: 0.0139 - val_acc: 0.9134
Epoch 3/25
60000/60000 [==============================] - 9s 149us/step - loss: 0.0128 - acc: 0.9205 - val_loss: 0.0114 - val_acc: 0.9278
Epoch 4/25
60000/60000 [==============================] - 8s 139us/step - loss: 0.0107 - acc: 0.9336 - val_loss: 0.0102 - val_acc: 0.9341
Epoch 5/25
60000/60000 [==============================] - 9s 148us/step - loss: 0.0094 - acc: 0.9410 - val_loss: 0.0094 - val_acc: 0.9407
Epoch 6/25
60000/60000 [==============================] - 9s 156us/step - loss: 0.0085 - acc: 0.9477 - val_loss: 0.0088 - val_acc: 0.9440
Epoch 7/25
60000/60000 [==============================] - 9s 155us/step - loss: 0.0078 - acc: 0.9523 - val_loss: 0.0082 - 

In [94]:
score = model.evaluate(x_test_1, y_test_1)
print('loss:', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 1s 95us/step
loss: 0.00561000154048088
正確率 0.9632


## 結果：
跟上次的結果比對是 **0.9632**  vs **0.9456**

這次改的東西：

1. "sigmoid"改成"relu"
2. 加了分叉
3. 改了epoch 和batch size